### ISSUE TITLE
Crate 2.0.6 Memory leak (Upto 3GB) and taking CPU 11 to 17%, Data missing due to memory leak in crate.

### ISSUE THREAD

**Issue by AbrarMalik** [crate/crate#6053]
**CrateDB version**:

**JVM version **:
JRE8 update 131

**OS / environment description**:
Windows Server 2012 R2, 8 GB RAM, 2.8 GHz

**Problem description**:

Integration objects provide us support for crate DB to log directly alarms data into crate DB by using Crate ODBC driver.
So IO AnE archiver integrate with Crate DB and 50 alarms per seconds are being log into crate DB. This test has been running from last friday. On Friday crate DB was running normally but i check today the crate DB memory leak observed on java instance. And crate is missing the records due to this memory raise.

```
[2017-07-31T00:00:03,981][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][136956][35612] duration [5.2s], collections [1]/[5.6s], total [5.2s]/[1d], memory [690.8mb]->[689mb]/[989.8mb], all_pools {[young] [21.1mb]->[18.8mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [669.7mb]->[670.2mb]/[682.6mb]}
[2017-07-31T00:00:03,988][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136956] overhead, spent [5.2s] collecting in the last [5.6s]
[2017-07-31T00:00:08,155][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136957] overhead, spent [3.5s] collecting in the last [4.1s]
[2017-07-31T00:00:12,715][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136958] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T00:00:16,752][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136959] overhead, spent [3.6s] collecting in the last [3.9s]
[2017-07-31T00:00:21,219][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136960] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T00:00:24,733][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136961] overhead, spent [3.1s] collecting in the last [3.5s]
[2017-07-31T00:00:29,179][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136962] overhead, spent [4s] collecting in the last [4.4s]
[2017-07-31T00:00:33,653][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136963] overhead, spent [4s] collecting in the last [4.3s]
[2017-07-31T00:00:38,041][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136964] overhead, spent [4s] collecting in the last [4.4s]
[2017-07-31T00:00:41,543][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136965] overhead, spent [3.1s] collecting in the last [3.4s]
[2017-07-31T00:00:47,250][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][136966][35622] duration [5.3s], collections [1]/[5.7s], total [5.3s]/[1d], memory [688.9mb]->[686.9mb]/[989.8mb], all_pools {[young] [20.3mb]->[18.8mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [670.1mb]->[668.1mb]/[682.6mb]}
[2017-07-31T00:00:47,250][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136966] overhead, spent [5.3s] collecting in the last [5.7s]
[2017-07-31T00:00:50,700][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136967] overhead, spent [3s] collecting in the last [3.4s]
[2017-07-31T00:00:55,199][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136968] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T00:00:58,675][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136969] overhead, spent [3s] collecting in the last [3.4s]
[2017-07-31T00:01:03,341][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][136970] overhead, spent [4.3s] collecting in the last [4.7s]
[2017-07-31T11:19:03,697][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146169] overhead, spent [4.8s] collecting in the last [5.1s]
[2017-07-31T11:19:10,743][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146170][44813] duration [6.6s], collections [1]/[7s], total [6.6s]/[1.4d], memory [704.2mb]->[701.7mb]/[989.8mb], all_pools {[young] [39.6mb]->[19.2mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [664.6mb]->[682.4mb]/[682.6mb]}
[2017-07-31T11:19:10,746][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146170] overhead, spent [6.6s] collecting in the last [7s]
[2017-07-31T11:19:16,096][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146171] overhead, spent [4.9s] collecting in the last [5.3s]
[2017-07-31T11:19:22,506][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146172][44815] duration [6s], collections [1]/[6.4s], total [6s]/[1.4d], memory [702.1mb]->[702.9mb]/[989.8mb], all_pools {[young] [38.5mb]->[38.4mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [663.6mb]->[664.4mb]/[682.6mb]}
[2017-07-31T11:19:22,525][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146172] overhead, spent [6s] collecting in the last [6.4s]
[2017-07-31T11:19:26,566][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146173] overhead, spent [3.5s] collecting in the last [3.8s]
[2017-07-31T11:19:31,227][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146174] overhead, spent [4.3s] collecting in the last [4.8s]
[2017-07-31T11:19:34,933][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146175] overhead, spent [3.3s] collecting in the last [3.7s]
[2017-07-31T11:19:40,123][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146176] overhead, spent [4.8s] collecting in the last [5.1s]
[2017-07-31T11:19:43,796][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146177] overhead, spent [3.2s] collecting in the last [3.5s]
[2017-07-31T11:19:48,498][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146178] overhead, spent [4.3s] collecting in the last [4.8s]
[2017-07-31T11:19:53,391][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146179] overhead, spent [4.5s] collecting in the last [4.8s]
[2017-07-31T11:19:58,869][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146180][44823] duration [5s], collections [1]/[5.4s], total [5s]/[1.4d], memory [703.2mb]->[703.4mb]/[989.8mb], all_pools {[young] [38.8mb]->[38.4mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [664.8mb]->[664.9mb]/[682.6mb]}
[2017-07-31T11:19:58,873][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146180] overhead, spent [5s] collecting in the last [5.4s]
[2017-07-31T11:20:03,971][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146181] overhead, spent [4.6s] collecting in the last [4.8s]
[2017-07-31T11:20:10,715][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146182][44825] duration [6.3s], collections [1]/[6.9s], total [6.3s]/[1.4d], memory [702.8mb]->[703.1mb]/[989.8mb], all_pools {[young] [39.8mb]->[38.4mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [664.3mb]->[664.6mb]/[682.6mb]}
[2017-07-31T11:20:10,721][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146182] overhead, spent [6.3s] collecting in the last [6.9s]
[2017-07-31T11:20:14,418][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146183] overhead, spent [3.3s] collecting in the last [3.7s]
[2017-07-31T11:20:19,096][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146184] overhead, spent [4.2s] collecting in the last [4.6s]
[2017-07-31T11:20:23,818][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146185] overhead, spent [4.2s] collecting in the last [4.5s]
[2017-07-31T11:20:28,198][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146186] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T11:20:31,776][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146187] overhead, spent [3.2s] collecting in the last [3.5s]
[2017-07-31T11:20:36,159][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146188] overhead, spent [3.9s] collecting in the last [4.3s]
[2017-07-31T11:20:41,036][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146189] overhead, spent [4.5s] collecting in the last [4.8s]
[2017-07-31T11:20:45,916][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146190] overhead, spent [4.4s] collecting in the last [4.8s]
[2017-07-31T11:20:49,414][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146191] overhead, spent [3.1s] collecting in the last [3.4s]
[2017-07-31T11:20:54,232][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146192] overhead, spent [4.3s] collecting in the last [4.8s]
[2017-07-31T11:20:58,394][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146193] overhead, spent [3.7s] collecting in the last [4.1s]
[2017-07-31T11:21:02,891][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146194] overhead, spent [4.1s] collecting in the last [4.4s]
[2017-07-31T11:21:06,960][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146195] overhead, spent [3.6s] collecting in the last [3.8s]
[2017-07-31T11:21:12,267][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146196] overhead, spent [4.9s] collecting in the last [5.4s]
[2017-07-31T11:21:15,855][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146197] overhead, spent [3.2s] collecting in the last [3.4s]
[2017-07-31T11:21:20,227][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146198] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T11:21:23,920][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146199] overhead, spent [3.1s] collecting in the last [3.5s]
[2017-07-31T11:21:28,437][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146200] overhead, spent [4.2s] collecting in the last [4.6s]
[2017-07-31T11:21:33,019][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146201] overhead, spent [4s] collecting in the last [4.4s]
[2017-07-31T11:21:37,463][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146202] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T11:21:41,039][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146203] overhead, spent [3.1s] collecting in the last [3.5s]
[2017-07-31T11:21:45,385][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146204] overhead, spent [3.9s] collecting in the last [4.3s]
[2017-07-31T11:21:49,076][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146205] overhead, spent [3.3s] collecting in the last [3.6s]
[2017-07-31T11:21:53,463][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146206] overhead, spent [4s] collecting in the last [4.3s]
[2017-07-31T11:21:57,975][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146207] overhead, spent [4.2s] collecting in the last [4.5s]
[2017-07-31T11:22:03,046][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146208] overhead, spent [4.5s] collecting in the last [5s]
[2017-07-31T11:22:06,640][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146209] overhead, spent [3.1s] collecting in the last [3.4s]
[2017-07-31T11:22:11,062][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146210] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T11:22:14,468][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146211] overhead, spent [3s] collecting in the last [3.4s]
[2017-07-31T11:22:19,065][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146212] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T11:22:23,111][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146213] overhead, spent [3.6s] collecting in the last [3.8s]
[2017-07-31T11:22:29,621][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146214][44857] duration [6s], collections [1]/[6.7s], total [6s]/[1.4d], memory [702.3mb]->[704.3mb]/[989.8mb], all_pools {[young] [39.9mb]->[38.8mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [663.8mb]->[665.5mb]/[682.6mb]}
[2017-07-31T11:22:29,637][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146214] overhead, spent [6s] collecting in the last [6.7s]
[2017-07-31T11:22:33,329][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146215] overhead, spent [3.3s] collecting in the last [3.7s]
[2017-07-31T11:22:38,232][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146216] overhead, spent [4.5s] collecting in the last [4.9s]
[2017-07-31T11:22:43,226][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146217] overhead, spent [4.5s] collecting in the last [4.7s]
[2017-07-31T11:22:49,780][INFO ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][old][146218][44861] duration [6.1s], collections [1]/[6.8s], total [6.1s]/[1.4d], memory [703.4mb]->[703.1mb]/[989.8mb], all_pools {[young] [40.4mb]->[38.4mb]/[273mb]}{[survivor] [0b]->[0b]/[34.1mb]}{[old] [664.9mb]->[664.6mb]/[682.6mb]}
[2017-07-31T11:22:49,793][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146218] overhead, spent [6.1s] collecting in the last [6.8s]
[2017-07-31T11:22:54,320][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146219] overhead, spent [4s] collecting in the last [4.3s]
[2017-07-31T11:22:58,670][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146220] overhead, spent [3.9s] collecting in the last [4.5s]
[2017-07-31T11:23:02,269][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146221] overhead, spent [3.1s] collecting in the last [3.4s]
[2017-07-31T11:23:06,693][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146222] overhead, spent [4s] collecting in the last [4.6s]
[2017-07-31T11:23:10,437][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146223] overhead, spent [3.3s] collecting in the last [3.6s]
[2017-07-31T11:23:14,830][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146224] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T11:23:18,430][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146225] overhead, spent [3.2s] collecting in the last [3.4s]
[2017-07-31T11:23:23,096][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146226] overhead, spent [4.2s] collecting in the last [4.8s]
[2017-07-31T11:23:26,985][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146227] overhead, spent [3.4s] collecting in the last [3.8s]
[2017-07-31T11:23:31,509][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146228] overhead, spent [4s] collecting in the last [4.5s]
[2017-07-31T11:23:34,938][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146229] overhead, spent [3s] collecting in the last [3.4s]
[2017-07-31T11:23:39,501][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146230] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T11:23:43,225][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146231] overhead, spent [3.3s] collecting in the last [3.7s]
[2017-07-31T11:23:47,681][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146232] overhead, spent [4s] collecting in the last [4.4s]
[2017-07-31T11:23:51,308][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146233] overhead, spent [3.2s] collecting in the last [3.4s]
[2017-07-31T11:23:55,764][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146234] overhead, spent [4.1s] collecting in the last [4.5s]
[2017-07-31T11:24:00,368][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146235] overhead, spent [4.1s] collecting in the last [4.6s]
[2017-07-31T11:24:05,640][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146236] overhead, spent [4.8s] collecting in the last [5.2s]
[2017-07-31T11:24:10,437][WARN ][o.e.m.j.JvmGcMonitorService] [Briceljk] [gc][146237] overhead, spent [4.4s] collecting in the last [4.7s]
```

**Steps to reproduce**:

1. AnE archiver subscribe with OPC Alarms server
2. Crate historian in AnE archiver with Crate DB by using ODBC driver
3.  Integrate crate DB with Integration Object AnE archiver.
4. Start archiver historian for logging OPC alarms into crate DB and let it for 5 to 6 hours, Crate java instance memory start raising. 

**Use case**:

**Feature description**:

**Comment by jodok** [comment]
invited!

**Comment by jodok** [comment]
sorry for closing and reopening :)

**Comment by seut** [comment]
@AbrarMalik 
Thanks for reporting we'll investigate this issue.
As this log entries are just kind a normal java garbage collection runs, I can't identifier this directly with your reported issues. But right, memory should not grow just by inserting data.
Please provide us a bit more information, like:
What is your table schema? most important, is it partitioned? do you use generated columns? How much shards/replicas are defined?
How much nodes are you running, do you monitor this on all nodes?
What kind of inserts are you doing (single insert, bulk insert, copy from, insert-by-subquery)?
Are you sure nothing was going on besides inserts (e.g. selects, opened admin-ui..)?
Can you explain us a bit more how you define that inserts are lost?

Sorry for that amount of questions, but this would help us a lot investigating.

**Comment by AbrarMalik** [comment]
@seut 
Thanks for your response.
I already mention in problem description that 50 records per seconds are being write on crate.
I am using only single node of crate.
 The following is the schema details.

CREATE TABLE IF NOT EXISTS "doc"."ams_data" (
   "ackreq" STRING,
   "activetime" TIMESTAMP,
   "activetime_ms" INTEGER,
   "actorid" STRING,
   "attributes" STRING,
   "conditions" STRING,
   "cookie" STRING,
   "eventcategory" STRING,
   "eventtime" TIMESTAMP,
   "eventtime_ms" INTEGER,
   "eventtype" STRING,
   "machinename" STRING,
   "mask" STRING,
   "message" STRING,
   "newstate" STRING,
   "quality" STRING,
   "servernodename" STRING,
   "serverprogid" STRING,
   "severity" INTEGER,
   "sourcename" STRING,
   "subcondition" STRING,
   "subscriptionname" STRING,
   PRIMARY KEY ("machinename", "serverprogid", "servernodename", "subscriptionname", "sourcename", "eventtime", "eventtime_ms", "severity", "message", "quality", "conditions", "subcondition", "mask", "newstate", "eventtype", "eventcategory", "ackreq", "activetime", "activetime_ms", "cookie", "actorid", "attributes")
)
CLUSTERED INTO 4 SHARDS
WITH (
   "blocks.metadata" = false,
   "blocks.read" = false,
   "blocks.read_only" = false,
   "blocks.write" = false,
   column_policy = 'dynamic',
   number_of_replicas = '0-1',
   "recovery.initial_shards" = 'quorum',
   refresh_interval = 1000,
   "routing.allocation.enable" = 'all',
   "routing.allocation.total_shards_per_node" = -1,
   "translog.durability" = 'REQUEST',
   "translog.flush_threshold_size" = 536870912,
   "translog.sync_interval" = 5000,
   "unassigned.node_left.delayed_timeout" = 60000,
   "warmer.enabled" = true,
   "write.wait_for_active_shards" = 'all'
)  

Please provide me the right settings for single node if i did something wrong with that because i am new for crate DB.
Java memory raise is very linear its start increasing when requests put on crate DB.

**Comment by seut** [comment]
@AbrarMalik We're still investigating and trying hard to reproduce.
Meanwhile, could you maybe try out the latest 2.1 (testing) release to see if it behaves the same?
We've done some improvements there already, so would be interesting if this issue is maybe already solved.
Also it would be very helpful if you could tell us how you are inserting your data (e.g. COPY FROM). Thanks.

**Comment by AbrarMalik** [comment]
@seut I have environment where the issue is still in produced form. You can connect with setup through teamviewer. Please let me know if you are interested.
I was using crate 1.2 but @Clause recommend me crate 2.0.6 version. I have setup too but the issue still exist.  Please take the remote of my system and investigate the issue. I am available 10 AM to 7 PM Pakistan Standard Time.

**Comment by joemoe** [comment]
hi @AbrarMalik, sorry, but we can't provide remote debugging for open source users. 
it would help us if you'd answer all of the questions @seut asked:

> ✅ What is your table schema? 
> ✅  most important, is it partitioned? 
> ✅ do you use generated columns?
> ✅  How much shards/replicas are defined?
> How much nodes are you running, do you monitor this on all nodes?
> What kind of inserts are you doing (single insert, bulk insert, copy from, insert-by-subquery)?
> Are you sure nothing was going on besides inserts (e.g. selects, opened admin-ui..)?
> Can you explain us a bit more how you define that inserts are lost?

**Comment by AbrarMalik** [comment]
@joemoe The following are the answer of the questions.

ü  What is your table schema?

IO: You can find below the table schema used for CrateDB:

   "ackreq" STRING,
   "activetime" TIMESTAMP,
   "activetime_ms" INTEGER,
   "actorid" STRING,
   "attributes" STRING,
   "conditions" STRING,
   "cookie" STRING,
   "eventcategory" STRING,
   "eventtime" TIMESTAMP,
   "eventtime_ms" INTEGER,
   "eventtype" STRING,
   "machinename" STRING,
   "mask" STRING,
   "message" STRING,
   "newstate" STRING,
   "quality" STRING,
   "servernodename" STRING,
   "serverprogid" STRING,
   "severity" INTEGER,
   "sourcename" STRING,
   "subcondition" STRING,
   "subscriptionname" STRING,

*   most important, is it partitioned?

IO: The table is not partitioned.

*  do you use generated columns?

IO: No.

*  How much shards/replicas are defined?

IO: No shards/replicas have been defined.

*  What kind of inserts are you doing (single insert, bulk insert, copy from, insert-by-subquery)?

IO: single insert

*  Are you sure nothing was going on besides inserts (e.g. selects, opened admin-ui..)?

IO: We are performing a database connection status checks in parallel with the insert queries.

**Comment by AbrarMalik** [comment]
@joemoe The following are the answer of the questions.

ü What is your table schema?

IO: You can find below the table schema used for CrateDB:

"ackreq" STRING,
"activetime" TIMESTAMP,
"activetime_ms" INTEGER,
"actorid" STRING,
"attributes" STRING,
"conditions" STRING,
"cookie" STRING,
"eventcategory" STRING,
"eventtime" TIMESTAMP,
"eventtime_ms" INTEGER,
"eventtype" STRING,
"machinename" STRING,
"mask" STRING,
"message" STRING,
"newstate" STRING,
"quality" STRING,
"servernodename" STRING,
"serverprogid" STRING,
"severity" INTEGER,
"sourcename" STRING,
"subcondition" STRING,
"subscriptionname" STRING,

most important, is it partitioned?
IO: The table is not partitioned.

do you use generated columns?
IO: No.

How much shards/replicas are defined?
IO: No shards/replicas have been defined.

What kind of inserts are you doing (single insert, bulk insert, copy from, insert-by-subquery)?
IO: single insert

Are you sure nothing was going on besides inserts (e.g. selects, opened admin-ui..)?
IO: We are performing a database connection status checks in parallel with the insert queries.

**Comment by seut** [comment]
@AbrarMalik ok thanks for the answers, just one is still missing ;) 

> Can you explain us a bit more how you define that inserts are lost?

Do you have any special configuration, especially did you enabled the stats by `stats.enabled=true`?
Also your PRIMARY KEY definition looks a bit odd to me as you include all columns defined. Is this really what you want? This will decrease insert and read performance, specially very fast realtime by-PK lookups will then only work if you provide all column in the where clause.
At all my tests, HEAP usage will increase by inserts (this is naturally as the data must be processed), but a garbage collector run will release it very soon. I cannot reproduce a constant increase in HEAP. Btw. what is your configured HEAP, how much memory does the machine has? Thanks.

**Comment by AbrarMalik** [comment]
@seut The following are the answer of your questions.

Do you have any special configuration, especially did you enabled the stats by stats.enables=true?
IO: No, we are using the default configuration.

Also your PRIMARY KEY definition looks at bit odd to me as you include all columns defined. Is this really what you want?

IO: PRIMARY KEY columns can be impact on memory raise? This not necessary to use all columns as PK.

This will decrease insert and read performance , especially very fast real time by PK lookups will then only work if you provide all column in the where clause.

At all my tests, HEAP usage will increase by insert (this is naturally as the data must be processed), but a garbage collector run will release it very soon. I cannot reproduce a constant increase in HEAP. Between what is you configured HEAP, how much memory does the machine has ?

IO: We did not configure the HEAP size and my machine have 8GB RAM. I could not find any memory decrease at all during the test run.

**Comment by seut** [comment]
@AbrarMalik  We've finally found a memory leak happening on single inserts. See https://github.com/crate/crate/commit/392572f8d7c5861fee178066c4aa1d73968c076c.
This will be released with the next hotfix release.
If it won't fix this issue, feel free to re-open it.
Thanks for reporting.

**Comment by AbrarMalik** [comment]
@seut Thanks for your support.
Please let me know that how can i incorporate these fix files into my crate setup. Please mention the procedure here.

**Comment by jodok** [comment]
the fixes will soon be on the testing release, and once they have stabilized in the stable release. you just need to wait a bit and take a look at the release notes.

**Comment by AbrarMalik** [comment]
@jodok Thanks for respond 
In which testing release did you incorporate fixture of this memory leak issue? Please let me know because i have to test it.

**Comment by AbrarMalik** [comment]
@jodok Can you please share the version?

**Comment by AbrarMalik** [comment]
@jodok @seut 
I am using crate 2.1.5 on windows server 2012 R2. Crate DB some exception during start.
When i start writing data on crate then traces would also start printing on command line, this cause to slow the DB response. Please suggest the solution for this. You can see attached log file for further details.  
[crate_logs.txt](https://github.com/crate/crate/files/1241667/crate_logs.txt)

**Comment by mxm** [comment]
@AbrarMalik Thank you for your feedback! Crate 2.1.6 will be released today or tomorrow. The release will get rid of the trace statements you were seeing. It will also contain a fix for the Windows startup problems.
